ProjFS.Mac/PrjFSLib/PrjFSLib.cpp

```diff
@@ -162,21 +165,36 @@ static mutex s_fileLocksMutex;
 
 PrjFS_Result PrjFS_RegisterForOfflineIO()
 {
-    s_kernelServiceOfflineWriterConnection = PrjFSService_ConnectToDriver(UserClientType_OfflineIO);
-    if (s_kernelServiceOfflineWriterConnection == IO_OBJECT_NULL)
+    mutex_lock lock(s_kernelServiceOfflineClientMutex);
+    
+    if (s_kernelServiceOfflineClientCount == 0)
     {
-        return PrjFS_Result_EDriverNotLoaded;
+        assert(s_kernelServiceOfflineWriterConnection == IO_OBJECT_NULL);
+        s_kernelServiceOfflineWriterConnection = PrjFSService_ConnectToDriver(UserClientType_OfflineIO);
+        if (s_kernelServiceOfflineWriterConnection == IO_OBJECT_NULL)
+        {
+            return PrjFS_Result_EDriverNotLoaded;
+        }
     }
+    
+    ++s_kernelServiceOfflineClientCount;
 
     return PrjFS_Result_Success;
 }
 
 PrjFS_Result PrjFS_UnregisterForOfflineIO()
 {
+    mutex_lock lock(s_kernelServiceOfflineClientMutex);
+    
+    assert(s_kernelServiceOfflineClientCount > 0);
     assert(s_kernelServiceOfflineWriterConnection != IO_OBJECT_NULL);
-
-    IOServiceClose(s_kernelServiceOfflineWriterConnection);
-    s_kernelServiceOfflineWriterConnection = IO_OBJECT_NULL;
+    --s_kernelServiceOfflineClientCount;
+    if (s_kernelServiceOfflineClientCount == 0)
+    {
+        IOServiceClose(s_kernelServiceOfflineWriterConnection);
+        s_kernelServiceOfflineWriterConnection = IO_OBJECT_NULL;
```

Comment:
The next time register is called, will have a problem?

Reply:
As the count has hit 0, the next register call will re-connect to the kext. So I think I've got this right - if you've worked out a sequence of events that might not be handled correctly, please elaborate!